### PR TITLE
benchmarking script

### PR DESCRIPTION
**Pull request by jhdark**
## Proposed changes

Performance benchmarking is a useful way to keep track of how updates to the code affect performance.

A benchmark script is proposed comparing the performance of a simulation of gas permeation in pure fenicsx and in festim. It can be run using;
```python3 test/benchmark.py```
If the performance of festim is more than 20% worse than the fenics script on average, the test will fail rasing an error:
```ValueError: festim is {diff:.1f}% slower than fenics, current acceptble threshold of 20%```

More sophisticated methods of continuous testing can be implemented in the future: https://github.com/ionelmc/pytest-benchmark.  

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [x] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: benchmark.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+    festim_times = []
+    for i in range(repetitions):
+        start = time.time()
+        test_permeation_problem()
+        festim_times.append(time.time() - start)
+    festim_time = np.mean(festim_times)
+
+    diff = (np.abs(fenics_time - festim_time) / ((fenics_time + festim_time) / 2)) * 100
+    if diff > 20:
```
Comment (RemDelaporteMathurin) [inline benchmark.py]:
```suggestion
    if diff > 10:
```

File: benchmark.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        raise ValueError(f"festim is {diff:.1f}% slower than fenics")
```
Comment (RemDelaporteMathurin) [inline benchmark.py]: It'd be nice to print it nonetheless
```suggestion
        raise ValueError(f"festim is {diff:.1f}% slower than fenics")
    else:
        print(f"festim is {diff:.1f}% slower than fenics")
```

File: benchmark.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+    relative_error = relative_error[
+        np.where(analytical_flux > 0.01 * np.max(analytical_flux))
+    ]
+    error = relative_error.mean()
+
+    assert error < 0.01
+
+
+def test_festim_vs_fenics_permeation_benchmark():
```
Comment (RemDelaporteMathurin) [inline benchmark.py]:
```suggestion
def test_festim_vs_fenics_permeation_benchmark():
    """Runs a problem with pure fenicsx and the same problem with FESTIM and
    raise ValueError if difference is too high"""
```

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
- move the benchmark script to test
- change threshold

File: benchmark.py (hunk 1 above)
Comment (jhdark) [reply benchmark.py]: unfortunately, festim is already 10-12% slower as is, maybe 15 instead then?

File: test/benchmark.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+    start = time.time()
+    test_permeation_problem(mesh_size=20001)
+    festim_time = time.time() - start
+
+    diff = (fenics_time - festim_time) / fenics_time
+    if diff < -0.1:
+        raise ValueError(
+            f"festim is {np.abs(diff):.1%} slower than fenics, current acceptable threshold of 10%"
+        )
```
Comment (RemDelaporteMathurin) [inline test/benchmark.py:196]:
```suggestion
    threshold = -0.1
    if diff < threshold:
        raise ValueError(
            f"festim is {np.abs(diff):.1%} slower than fenics, current acceptable threshold of {threshold:.1%}"
        )
```

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
One small refactoring change and then good to merge!

**Pull request merged**